janus_server/src/bin/aggregation_job_driver.rs

```diff
@@ -0,0 +1,1422 @@
+                        report_aggregation.state = ReportAggregationState::Invalid;
+                    }
+                }
+
+                TransitionTypeSpecificData::Failed { error } => {
+                    // If the helper failed, we move to FAILED immediately.
+                    // TODO(brandon): is it correct to just record the transition error that the helper reports?
+                    error!(task_id = %task.id, aggregation_job_id = %aggregation_job.aggregation_job_id,
+                        report_nonce = %report_aggregation.nonce, helper_err = ?error, "Helper couldn't step report aggregation");
+                    report_aggregation.state = ReportAggregationState::Failed(error);
+                }
+            }
+            report_aggregations_to_write.push(report_aggregation);
+        }
+
+        // Determine if we've finished the aggregation job (i.e. if all report aggregations are in
+        // a terminal state), then write everything back to storage.
+        let aggregation_job_id = aggregation_job.aggregation_job_id;
+        let aggregation_job_is_finished = report_aggregations_to_write
+            .iter()
+            .all(|ra| !matches!(ra.state, ReportAggregationState::Waiting(_, _)));
+        let aggregation_job_to_write = if aggregation_job_is_finished {
+            let mut aggregation_job = aggregation_job;
+            aggregation_job.state = AggregationJobState::Finished;
+            Some(aggregation_job)
+        } else {
+            None
+        };
+        let report_aggregations_to_write = Arc::new(report_aggregations_to_write);
+        let aggregation_job_to_write = Arc::new(aggregation_job_to_write);
+        self.datastore
+            .run_tx(|tx| {
+                let (report_aggregations_to_write, aggregation_job_to_write) = (
+                    Arc::clone(&report_aggregations_to_write),
+                    Arc::clone(&aggregation_job_to_write),
+                );
+                Box::pin(async move {
+                    let report_aggregations_future =
+                        try_join_all(report_aggregations_to_write.iter().map(
+                            |report_aggregation| tx.update_report_aggregation(report_aggregation),
+                        ));
+                    let aggregation_job_future = try_join_all(
+                        aggregation_job_to_write
+                            .iter()
+                            .map(|aggregation_job| tx.update_aggregation_job(aggregation_job)),
+                    );
+
+                    try_join!(
+                        tx.release_aggregation_job(task.id, aggregation_job_id),
+                        report_aggregations_future,
+                        aggregation_job_future
+                    )?;
+                    Ok(())
+                })
+            })
+            .await?;
+        Ok(())
```

Comment:
BTW (and apologies for harping on this so much), the reason I'm so worried about aggregation idempotency: if there is a failure at almost any point in this function (more precisely, from about when we send the request to the helper, to when the datastore transaction to write our new state completes), the aggregation job will be in a "skewed" state between the leader & helper and will effectively be lost.

This failure might be a bug in our program logic, an OOM, a hardware failure in the VM's underlying hypervisor, etc.

While we might be able to shift things around to shrink the size of this "risk window", unless/until we can somehow make the helper's durable state update transaction with our own durable state update, the window will exist.

Reply:
(Another approach is noticing we are stuck/have failed & retrying the entire aggregation job from the start; but I think this would require some spec-level changes[1] as well as adding a fair bit of complexity to the aggregation job creator. Anyway, this is pretty far outside the scope of this PR.)

[1] Input Share Validation requires us to "Check if the report has already been aggregated." Right now the right way to interpret this is "report is part of an aggregation job"--if we interpret it as "report is part of a _successful, complete_ aggregation job" then we permit reports being aggregated multiple times since validation occurs only at the start of the aggregation process. Se we'd need some way for the leader/helper to agree that an aggregation job is canceled, or perhaps we'd want to update the checks to only allow finishing an aggregation once for a given report.